### ISSUE TITLE
jax.nn.tanh gradient returns zero for saturated float32 input

### ISSUE THREAD

**Issue by ALinrunrun** [jax-ml/jax#40388]
### Description

`jax.nn.tanh` appears to return an incorrect zero first-order derivative for a finite `float32` input.

For tanh, the derivative can be written as:

```
tanh'(x) = 4 * exp(-2 * abs(x)) / (1 + exp(-2 * abs(x)))**2
```

At `x = -9.010913848876953`, the high-precision reference rounded to `float32` is approximately `5.9604584379258085e-08`, which is finite, normal, and nonzero. However, JAX returns `0.0`.

The same issue is also relevant to the mapped public paths `jax.lax.tanh` and `jax.numpy.tanh`.

### Minimal reproducible example

```
#!/usr/bin/env python3
import os
import sys
import math

os.environ.setdefault("CUDA_VISIBLE_DEVICES", "-1")
os.environ.setdefault("JAX_PLATFORMS", "cpu")
os.environ.setdefault("TF_CPP_MIN_LOG_LEVEL", "3")

try:
    import jax
    import jax.numpy as jnp
except ImportError as e:
    print(f"missing dep: {e}")
    sys.exit(2)

jax.config.update("jax_enable_x64", True)

x = -9.010913848876953
expected = 5.9604584379258085e-08

actual = float(jax.grad(lambda z: jnp.sum(jax.nn.tanh(z)))(
    jnp.asarray(x, dtype=jnp.float32)
))

print("jax:", actual)
print("expected:", expected)

if not (math.isfinite(actual) and abs(actual - expected) <= 1e-6 * abs(expected)):
    sys.exit(0)
sys.exit(1)
```

### Actual result

```
jax: 0.0
expected: 5.9604584379258085e-08
```

### Expected result

```
5.9604584379258085e-08
```

The expected derivative is finite, normal, nonzero, and representable in `float32`. JAX instead returns zero, suggesting that the autodiff path loses a small but representable derivative signal when the forward `tanh` value rounds to `-1.0`.

### System info (python version, jaxlib version, accelerator, etc.)

jax: 0.10.1
jaxlib: 0.10.1
numpy: 2.4.6
python: 3.13.13
platform: Ubuntu 24.04.3 LTS, Linux x86_64
accelerator: CPU backend

Note: GPU is present, but CUDA-enabled jaxlib is not installed, so JAX falls back to CPU.

**Comment by jakevdp** [comment]
This falls under #38651 – issues like this (typically discovered via AI fuzzing of extreme input values) are low priority for the JAX team.
